include/gadget/filter.h

```diff
@@ -0,0 +1,128 @@
+/* SPDX-License-Identifier: (GPL-2.0 WITH Linux-syscall-note) OR Apache-2.0 */
+
+#ifndef FILTER_H
+#define FILTER_H
+
+#include <bpf/bpf_helpers.h>
+#include <gadget/macros.h>
+#include <gadget/types.h>
+#include <gadget/mntns_filter.h>
+
+#define GADGET_INVALID_ID ((uid_t)-1)
+
+// user space terminology
+const volatile gadget_pid targ_pid = 0;
+GADGET_PARAM(targ_pid);
+
+const volatile gadget_tid targ_tid = 0;
+GADGET_PARAM(targ_tid);
+
+const volatile gadget_uid targ_uid = GADGET_INVALID_ID;
+GADGET_PARAM(targ_uid);
+
+const volatile gadget_gid targ_gid = GADGET_INVALID_ID;
+GADGET_PARAM(targ_gid);
+
+const volatile gadget_comm targ_comm[TASK_COMM_LEN] = {};
+GADGET_PARAM(targ_comm);
+
+static __always_inline bool gadget_should_discard_pid_tid(gadget_pid pid,
+							  gadget_tid tid)
+{
+	if (targ_pid != 0 && targ_pid != pid)
+		return true;
+
+	if (targ_tid != 0 && targ_tid != tid)
+		return true;
+
+	return false;
+}
+
+static __always_inline bool gadget_should_discard_uid_gid(gadget_uid uid,
+							  gadget_gid gid)
+{
+	if (targ_uid != GADGET_INVALID_ID && targ_uid != uid)
+		return true;
+
+	if (targ_gid != GADGET_INVALID_ID && targ_gid != gid)
+		return true;
+
+	return false;
+}
+
+static __always_inline bool
+gadget_should_discard_comm(gadget_comm comm[TASK_COMM_LEN])
+{
+	if (targ_comm[0] == '\0')
+		return false;
+
+	for (int i = 0; i < TASK_COMM_LEN; i++) {
+		if (comm[i] != targ_comm[i])
+			return true;
+	}
+
+	return false;
+}
+
+// gadget_should_discard returns true if the gadget should skip this event.
+static __always_inline bool
+gadget_should_discard(gadget_mntns_id mntns_id, gadget_pid pid, gadget_tid tid,
+		      gadget_comm comm[TASK_COMM_LEN], gadget_uid uid,
+		      gadget_uid gid)
+{
+	if (gadget_should_discard_mntns_id(mntns_id))
+		return true;
+
+	if (gadget_should_discard_pid_tid(pid, tid))
+		return true;
+
+	if (gadget_should_discard_uid_gid(uid, gid))
+		return true;
+
+	if (gadget_should_discard_comm(comm))
+		return true;
+
+	return false;
+}
+
+// gadget_should_discard_current returns true if the gadget should drop this
+// event. This function uses the current mount namespace, pid, tid, uid, and gid
+// to determine if the event should be dropped.
+// This function is implemented in a way that it returns as soon as possible to
+// avoid unnecesary operations, hence there is some code duplication with the
+// gadget_should_discard function.
+static __always_inline bool gadget_should_discard_current()
```

Comment:
> it is about event.

I don't want to put event there as it's more generic, for instance snapshotters and other gadgets aren't about events but data.

Do you have any recommendation of how we could call this?

Reply:
Renamed them to `gadget_should_discard_data` and `gadget_should_discard_data_current`. Is it better?

Reply:
Maybe inverse? `gadget_relevant_env()` / `gadget_relevant_pid()` (or `_filter_` instead of `_relevant`) etc.? I also find the open "_current" confusing.

Reply:
> Maybe inverse? gadget_relevant_env() / gadget_relevant_pid() (or _filter_ instead of _relevant) etc.? I

I'd prefer to stick to `gadget_should_discard_` to keep it uniform with the existing `gadget_should_discard_mntns_id`. 

> I also find the open "_current" confusing.

According to my understanding "current" in the Linux kernel refers to the current task (https://stackoverflow.com/a/12434702). I tried to use the same terminology as it does the same, drop the event based on the current task. ebpf helpers are also named `bpf_get_current_pid_tgid` and `bpf_get_current_uid_gid`, `bpf_get_current_comm`. Any other idea to name this function?

Reply:
Explanation makes sense - I still don't like it 😄 

I suggested the change because the `_should_discard_X_(by_Y)` refers to whatever is being created (X) that now should be discarded (an assumption on the action) while a `_relevant_Y` would only refer to the actual data Y that is being checked, without an assumption on the action or X - making it IMHO easier to understand. In theory, you could also just use the functions to check for a targeted namespace and then do something else instead of discarding.

But yeah, this would mean we also have to change the `gadget_should_discard_mntns_id` to clean it up. Maybe just leave it for now and rethink later on.

Reply:
Just for the sake of completeness: `_targeted_` (or `_is_targeted_`) would align even better - so `gadget_targeted_env()`, `gadget_targeted_mntns_id()`, etc.

Reply:
> So, should we rename everything to have current too?

I think it's a good compromise. Some clarifications about the current status:
- There are gadget_should_discard_{pid,tid,uid,gid,comm,mntnsid} functions that receive as argument the value to discard. They don't need to have current in their name
- There is a `gadget_should_discard_data` that receives all fields. It's useful for gadgets like snapshot process that gather this information without using the `bpf_get_current_` helpers
- gadget_should_discard_data_current: Useful for discarding data based on current task. 

So I think we're aligned there. A missing piece is to create an alias for `gadget_get_current_mntns_id`. 

> Then, the only thing with gadget_should_discard_current() is the documentation should highlight this will discard the actual even generated by the current task struct, not all the events coming from the current task struct.

Documentation is: 

### `gadget_should_discard_data_current`

```c
bool gadget_should_discard_data_current()
```

This function returns `true` if an event should be discarded based on the
current process.

Do you think it should be changed in any way?

Reply:
It was added in the last commit of this PR.